### ISSUE TITLE
[Bug] SERVE_BATCH + COLI_METAL=1: responses are 1 token or garbage — Metal fused decode paths ignore ragged kvs[]/positions[]

### ISSUE THREAD

**Issue by TheDefend84** [JustVugg/colibri#285]
## Symptom

macOS, Apple Silicon (Mac Studio M4 Max, 64 GB), GLM-5.2 int4, engine built with `METAL=1`, running `COLI_METAL=1 coli serve`. Every `/v1/chat/completions` request returns either exactly **one token** followed by `finish_reason: "stop"` (e.g. `"To"` for "Compare RAM and VRAM caching"), or, depending on sampling, a longer stream of garbage tokens (`"To`20oor a72----- ... mine mine mine ..."`).

The same build answers perfectly via `coli run` / `coli chat`. The same serve setup answers perfectly **without** `COLI_METAL=1`. So: engine OK, server OK, Metal OK — only the *combination* serve+Metal is broken.

## Cause

`run_serve_mux` decodes through `step_decode_batch`, which forwards ragged rows: per-row `kvs[]` / `positions[]`, `pos_base=0`. The comment on `attention_rows` documents the contract:

> kvs/pos describe a ragged decode batch: each row may belong to a different sequence. NULL keeps the original contiguous, currently-bound KV path.

Both Metal fast paths check `S<=4`, dims, `m->kv_start[layer]==0` — but **not `kvs == NULL`**:

1. `attention_rows` (fused decode attention, `coli_metal_attn_decode`) — passes `m->Lc[layer]`, `m->Rc[layer]`, `pos_base`, `m->kv_start[layer]`, i.e. the *bound* KV state and *contiguous* positions, ignoring the per-row arrays.
2. `layer_forward_rows` (full-layer command buffer, `coli_metal_layer_decode`) — same pattern, same problem, and it runs *before* attention_rows, so guarding only (1) still yields garbage.

With `pos_base=0` the GPU kernel attends at position 0 instead of the end of the context and writes KV rows to the wrong place → near-random logits → the sampler frequently draws a stop token immediately (hence the 1-token replies; the first token is always fine because prefill still goes through the contiguous `step()` path).

The first decode step after prefill happens inside `mux_submit` via `pick_tok` on prefill logits (correct); every subsequent token comes from the broken batched forward.

## Fix (tested)

Add the missing ragged-batch check to both gates, exactly as the contract comment prescribes:

```c
/* attention_rows */
if(g_metal_enabled && !kvs && S<=4 && ... )

/* layer_forward_rows, full-layer CB */
if(g_metal_enabled && !kvs && S<=4 && li<c->n_layers && l->sparse && ... )
```

Ragged batches then take the CPU absorb path (which handles kvs[]/positions[] per row); routed experts stay on the GPU (`[METAL] mode: batched routed experts on GPU` still active).

Verified end-to-end on the machine above: serve+Metal now streams coherent, correctly formatted answers. `coli run`/`chat` performance is unaffected (still hits the fused paths).

A follow-up could teach the Metal kernels to take per-row KV/positions so serve mode regains fused attention on GPU — the guard is the minimal correctness fix.

## Important follow-up: the bug poisons `.coli_kv` persistently

While the unguarded Metal paths are active, the misplaced KV writes end up in the KV rows that `mux_done`/`kv_disk_append` persist to `<model>/.coli_kv` (KVSAVE=1 default). Every later serve start then *resumes* the corrupted state ("[KV] resumed conversation from disk"), so the corruption **survives restarts and even engine rebuilds** — outputs show doubled tokens ("the the", "optimism**ism**"), degrading into repetition loops, and greedy decoding becomes non-reproducible across otherwise identical requests. Anyone who ever ran serve+Metal on the affected code should delete `<model>/.coli_kv*` after upgrading, otherwise the fix appears not to work. A `.coli_kv` integrity/version check (or disabling KV persistence in the multiplexed server) would make this failure mode much less sticky.

## Repro

```
COLI_METAL=1 ./coli serve --model <glm52_i4> --ram 48 --port 8000
curl -s -X POST localhost:8000/v1/chat/completions -H 'Content-Type: application/json' \
  -d '{"model":"glm-5.2","messages":[{"role":"user","content":"Compare RAM and VRAM caching"}],"max_tokens":48}'
# -> {"content":"To","finish_reason":"stop","completion_tokens":1}
```

Happy to open a PR with the two-line patch.

**Comment by JustVugg** [comment]
Fixed on `dev` by #288 — same root cause you diagnosed: the fused decode kernels ignored ragged `kvs[]`/`positions[]` and are now gated off (`!kvs`) for the multiplexed serve path. Spot-on report, thank you.
